### PR TITLE
feat-21: Detect obstacles with ultrasonic sensors

### PR DESCRIPTION
**Pull request by kekoleko123**
It makes a publisher to notice when an obstacles is detected.

### CODE REVIEW

**Review by BloodFutur — COMMENTED** [review]
Great job, still the tests and documentation to do! :)

File: raspberryPI3/ros2_ws/src/detection/detection/obstacles_detection.py [hunk 1]
```diff
@@ -0,0 +1,44 @@
...
+from interfaces.msg import Ultrasonic
+
+class ObstaclesDetection(Node):
+    
+    def __init__(self):
+        super().__init__('obstacles_detection')
+        self.publisher = self.create_publisher(Bool, 'obstacles_detection', 10)
+        self.get_logger().info('obstacles_detection_node started')
+        self.subscriber = self.create_subscription(Ultrasonic ,'us_data' , self.obstacle_detected, 10)
+        self.detection_distance=50
```
Comment (BloodFutur) [inline raspberryPI3/ros2_ws/src/detection/detection/obstacles_detection.py:14]: You can use ROS parameters with something like this :
`self.detection_distance = self.declare_parameter('detection_distance', 50).get_parameter_value().integer_value`

File: raspberryPI3/ros2_ws/src/detection/detection/obstacles_detection.py [hunk 2]
```diff
@@ -0,0 +1,44 @@
+import rclpy
+from rclpy.node import Node
+
+from std_msgs.msg import Bool
+from interfaces.msg import Ultrasonic
+
+class ObstaclesDetection(Node):
+
```
Comment (BloodFutur) [inline raspberryPI3/ros2_ws/src/detection/detection/obstacles_detection.py:8]: You can add Docstring (IA does this very well), short example:

```py
class ObstaclesDetection(Node):
    """
    Node to detect obstacles using ultrasonic sensors.
    Publishes a Bool message to 'obstacles_detection' topic when an obstacle is detected.
    """
    ```

**Review by BloodFutur — COMMENTED** [review]
Try to run `colcon test` on your package (check docs online for the parameters to test only your package). It will use linters to check coding rules (flake8, pep257)

File: raspberryPI3/ros2_ws/src/detection/detection/obstacles_detection.py [hunk 3]
```diff
@@ -0,0 +1,44 @@
...
+        self.get_logger().info('obstacles_detection_node started')
+        self.subscriber = self.create_subscription(Ultrasonic ,'us_data' , self.obstacle_detected, 10)
+        self.detection_distance=50 
+                
+    def obstacle_detected(self, msg_us: Ultrasonic):
+        # We check the two sensors each because they don't look at the same place
+        # ex : if obstacle only at the left side, the right sensor won't see it
+        # We only send a true msg if we detect an obstacle so we don't need a false case
+        # We do that so it's easier for the future subscribers of the /obstacles_detection subscribers to process the msg
+        # And bc we don't need to change anything in the behavior of the car if there is no car
```
[the remark this answers is not in the document]
Comment (kekoleko123) [reply raspberryPI3/ros2_ws/src/detection/detection/obstacles_detection.py:21]: Thanks !

File: raspberryPI3/ros2_ws/src/detection/detection/obstacles_detection.py (hunk 1 above)
[the remark this answers is not in the document]
Comment (kekoleko123) [reply raspberryPI3/ros2_ws/src/detection/detection/obstacles_detection.py:14]: Done !

**Comment by kekoleko123** [comment]
Ok so i wrote ```colcon test --ctest-args tests detection``` which starts a lot of different tests on every packages (not only detection), then i wrote ``colcon test-results --all` and i have this : 
image

image

Don't even know what does it mean (or the difference between errors and failure).

File: raspberryPI3/ros2_ws/src/detection/detection/obstacles_detection.py (hunk 1 above)
Comment (kekoleko123) [reply raspberryPI3/ros2_ws/src/detection/detection/obstacles_detection.py:14]: Perfect  : 
image

File: raspberryPI3/ros2_ws/src/detection/detection/obstacles_detection.py (hunk 2 above)
Comment (kekoleko123) [reply raspberryPI3/ros2_ws/src/detection/detection/obstacles_detection.py:8]: Done !

File: raspberryPI3/ros2_ws/src/detection/detection/obstacles_detection.py (hunk 1 above)
Comment (BloodFutur) [reply raspberryPI3/ros2_ws/src/detection/detection/obstacles_detection.py:14]: Great

**Comment by BloodFutur** [comment]
> Ok so i wrote `colcon test --ctest-args tests detection` which starts a lot of different tests on every packages (not only detection), then i wrote ``colcon test-results --all` and i have this : image
> 
> imageon
> 
> Don't even know what does it mean (or the difference between errors and failure).

Try this command instead : `colcon test --event-handlers console_direct+ --packages-select detection`

It tests your package and display outputs in the terminal. You should see errors but that's OK, it is for you to understand what's going on.

The command you use says: "test all packages and for c++ tests add a parameter named 'detection'.

**Comment by kekoleko123** [comment]
Ok so i see many style errors/warnings, do i have to correct them all ?

**Comment by BloodFutur** [comment]
There are tools that can fix them for you automatically. I will check that. 
Then, you can make tests for your code. Create a script in the test folder, use pytest and test what your code do, we can see this together if you are struggling.

File: .devcontainer/devcontainer.json
Comment (BloodFutur) [inline .devcontainer/devcontainer.json:1]: Nice for linting and formatting.

File: raspberryPI3/ros2_ws/src/detection/Readme.md
Comment (BloodFutur) [inline raspberryPI3/ros2_ws/src/detection/Readme.md:1]: Nice README

File: raspberryPI3/ros2_ws/src/detection/test/test_obstacles_detection_wip.py
Comment (BloodFutur) [inline raspberryPI3/ros2_ws/src/detection/test/test_obstacles_detection_wip.py:1]: I do not know if it is a good idea to have this there. But at least the work is not lost and does not interfere with other code as it is commented. It seems ok for me.

File: raspberryPI3/ros2_ws/src/detection/test/test_unit_obstacles_detection.py
Comment (BloodFutur) [inline raspberryPI3/ros2_ws/src/detection/test/test_unit_obstacles_detection.py:1]: Very good tests

**Review by BloodFutur — APPROVED** [review]
Very good.

**Pull request merged**